### PR TITLE
[LO extension / Standalone] add the possibility to define more than o…

### PR DESCRIPTION
**Pull request by FredKruse**
The change creates the ability to define more than one custom option per rule.
All existing rules with custom options have been adjusted. The options panel has been expanded accordingly.
The API and the HTTP server (ConfigurationInfo) have been adapted. Upward and downward compatibility between versions lower than 6.5 and higher than 6.4 was ensured.
A second option was added for the GermanFillerWordsRule as a test case.
All changes were tested.
To-do: Some more rules should be expanded to include additional options.
A merge can be performed.

### CODE REVIEW

File: languagetool-gui-commons/src/main/java/org/languagetool/gui/Configuration.java [hunk 1]
```diff
@@ -1466,13 +1466,38 @@ private void parseUnderlineTypes(String typessString, Map<String, Short> underli
...
+          if (ruleAndValue.length != 2) {
+            throw new RuntimeException("Could not parse rule and value, colon expected: '" + ruleToValue + "'");
+          }
+          String[] values = ruleAndValue[1].split(";");
+          Object[] objects = new Object[values.length];
+          for (int i = 0; i < values.length; i++) {
+            String str = values[i].trim();
+            char c = values[i].charAt(0);
+            str = str.substring(1);
+            if (c == 's') {
```
Comment (danielnaber) [inline languagetool-gui-commons/src/main/java/org/languagetool/gui/Configuration.java:1483]: A comment with examples would be nice here

File: languagetool-core/src/main/java/org/languagetool/rules/RuleOption.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+   * Get the maximum of a configurable value
+   */
+  public Object getMaxConfigurableValue() {
+    return maxConfigurableValue;
+  }
+
+
+
+  
+
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/RuleOption.java:80]: empty lines

File: languagetool-office-extension/src/main/java/org/languagetool/openoffice/LORemoteLanguageTool.java [hunk 3]
```diff
@@ -309,12 +310,37 @@ public void disableCategory(CategoryId id) {
...
     for (String rule : rules) {
-      String ruleValueString = rule + ":" + configurableValues.get(rule);
-      ruleValues.add(ruleValueString);
+      Object[] obs = configurableValues.get(rule);
+      if (obs != null && obs.length > 0) {
+        StringBuilder sbOV = new StringBuilder();
+        char c;
+        for (int i = 0; i < obs.length; i++) {
+          Object o = obs[i];
+          if (o instanceof Integer) {
```
Comment (danielnaber) [inline languagetool-office-extension/src/main/java/org/languagetool/openoffice/LORemoteLanguageTool.java:323]: Is this code duplicated from above? Can we avoid the duplication?

File: languagetool-office-extension/src/main/java/org/languagetool/openoffice/LORemoteLanguageTool.java [hunk 4]
```diff
@@ -478,6 +505,48 @@ static class RemoteRule extends Rule {
...
+        ruleOptions = new RuleOption[rOptions.size()];
+        for (int i = 0; i < rOptions.size(); i++) {
+          String defaultType = (String) rOptions.get(i).get("defaultType");
+          String configureText = (String) rOptions.get(i).get("configureText");
+          Object defaultValue;
+          Object minConfigurableValue;
+          Object maxConfigurableValue;
+          if (defaultType.equals("Integer")) {
+            defaultValue = (int) rOptions.get(i).get("defaultValue");
+            minConfigurableValue = (int) rOptions.get(i).get("minConfigurableValue");
```
Comment (danielnaber) [inline languagetool-office-extension/src/main/java/org/languagetool/openoffice/LORemoteLanguageTool.java:521]: `minConfigurableValue` etc. should probably be a constant, as it is used several times

File: languagetool-office-extension/src/main/java/org/languagetool/openoffice/LORemoteLanguageTool.java [hunk 5]
```diff
@@ -590,6 +648,47 @@ static class RemoteTextLevelRule extends TextLevelRule {
         setOfficeDefaultOff();
       }
       isDictionaryBasedSpellingRule = ruleMap.containsKey("isDictionaryBasedSpellingRule");
+      if (ruleMap.containsKey("ruleOptions")) {
+        Object o = ruleMap.get("ruleOptions");
+        @SuppressWarnings("unchecked")
+        List<Map<String,Object>> rOptions = (List<Map<String, Object>>) o;
+        ruleOptions = new RuleOption[rOptions.size()];
+        for (int i = 0; i < rOptions.size(); i++) {
```
Comment (danielnaber) [inline languagetool-office-extension/src/main/java/org/languagetool/openoffice/LORemoteLanguageTool.java:656]: This code is duplicated?

File: languagetool-server/src/main/java/org/languagetool/server/TextChecker.java [hunk 6]
```diff
@@ -732,16 +732,37 @@ private Map<String, Integer> getRuleMatchCount(List<CheckResults> res) {
...
+        } else if (c == 'f') {
+          objects[i] = Float.parseFloat(str);
+        } else if (c == 'd') {
+          objects[i] = Double.parseDouble(str);
+        } else if (c == 'c') {
+          objects[i] = str.charAt(0);
+        } else if (c == 'i') {
+          objects[i] = Integer.parseInt(str);
+        } else {  //  compatible to old version 
+          objects[i] = Integer.parseInt(values[i]);
```
Comment (danielnaber) [inline languagetool-server/src/main/java/org/languagetool/server/TextChecker.java:762]: This is duplicated?

**Review by danielnaber — CHANGES_REQUESTED** [review]

**Comment by danielnaber** [comment]
@fabrichter Maybe you can quickly check whether the change in `UserConfig` is harmless?

**Comment by FredKruse** [comment]
The requested changes was done

**Comment by fabrichter** [comment]
@danielnaber 
For UserConfig, the change to an Array probably breaks `.equals` and thus pipeline caching because equals on arrays checks for reference equality. But this most likely isn't relevant in practice, since for the use case where pipeline caching is enabled, configurableRuleValues is always empty. So I think we don't need to change this, but a comment about it in `.equals` would be good to add.

I would also add a better getter for the configuration values like this:
```java

  public <T> T getConfigValueByID(String ruleID, int index, Class<T> clazz, T defaultValue) {
    Object value = configurableRuleValues.get(ruleID);
    if (value == null) return defaultValue;

    if (index >= value.length) return defaultValue;
    if (!clazz.isInstance(value[index])) return defaultValue;

    return (T) value;
  }

```

This should be used instead of the existing manual accesses and checks to inner values of `configurableRuleValues`, as the checks are repetitive and not done consistently as-is.

**Comment by FredKruse** [comment]
The requested changes from @fabrichter and @danielnaber was done.

**Pull request merged**